ui/components/app/signature-request/signature-request.component.js

```diff
@@ -45,9 +45,85 @@ export default class SignatureRequest extends PureComponent {
       hardwareWalletRequiresConnection,
     } = this.props;
     const { address: fromAddress } = fromAccount;
-    const { message, domain = {} } = JSON.parse(data);
+    const { message, domain = {}, types, primaryType } = JSON.parse(data);
     const { metricsEvent } = this.context;
 
+    const mapType = (msgType, definedType) => {
+      // try and map it to a solidity type
+      if (
+        (definedType.type.indexOf('int') >= 0 ||
+          definedType.type.indexOf('fixed') >= 0) &&
+        msgType === 'number'
+      ) {
+        return true;
+      } else if (
+        (definedType.type.indexOf('bytes') >= 0 ||
+          definedType.type === 'address') &&
+        msgType === 'string'
+      ) {
+        return true;
+      }
+
+      return false;
+    };
+
+    const sanitizeMessage = (msg, baseType) => {
+      if (version === 'V4') {
```

Comment:
Why only V4?

Reply:
I thought only V4 has types...just checked test-app and saw it not.